enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,271 @@
+    bundles to be refreshed.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Generate an intermediate CA certificate with the same public key as the current CA
+  but signed with the private key of the new CA.
+  - This intermediate certificate should be included in the the new CA
+    bundle along with the new CA's primary certificate, and serves to
+    bridge trust between an unrefreshed server (serving with a cert
+    generated from the previous CA) and a refreshed client (validating
+    trust with the new CA bundle).
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+  - In addition to ensuring regeneration when `AuthorityKeyId` and `SubjectKeyId`
+    are set, this comparison schema will also ensure regeneration of serving
+    certs without `SubjectKeyId` set (due to being generated by a CA without
+    `AuthorityKeyId` set) since an empty `SubjectKeyId` will not match a
+    populated `AuthorityKeyId`.
+
+- Provide an indication that rotation has taken place via an event.
+
+- Allow rotation to be forced when
+  `ServiceCA.Spec.ForceRotation.Reason` is set to a non-empty value
+  that does not match the value for the
+  `service.beta.openshift.io/forced-rotation-reason` annotation on the
+  signing secret.
+  - Setting an annotation on the signing secret ensures that a forced
+    rotation is triggered at most once per reason.
+  - If a forced rotation and expiry-triggered rotation coincide, only
+    the forced rotation should be performed.
+
+- Allow the duration of the grace period for continuing to trust the
+  previous CA to be overridden when forcing rotation by setting
+  `ServiceCA.Spec.ForceRotation.GraceDuration`.
+  - If set, the expiry of the intermediate certs bridging trust
+    between the pre-rotation serving certs and CA bundle will be set
+    to the time of rotation plus the provided duration.
+
+- Allow trust to be maintained across multiple rotations by setting
```

Comment:
@mrogers950 I really appreciate your thoroughness in reviewing this enhancement!

The item starting here is new - I added it in response to a problem I ran into in testing. Since trust is only maintained for the current and previous CAs, more than one rotation can result in broken trust if the rotations happen quickly enough. Since I'd like to test both forced and time-based rotation in e2e, I think it would be useful to allow trust to be maintained between clients and servers that have not refreshed, have refreshed after the first rotation, and that have refreshed after the second rotation. WDYT?

Reply:
e2e for forced rotation [is passing](https://github.com/openshift/service-ca-operator/pull/77) without this change. I think this feature might still be useful for exploratory testing, but maybe it should be relegated to `unsupportedConfigOverrides`?

cc: @sttts